backend/hardware/devicelist.py

```diff
@@ -55,11 +72,18 @@ def setupDevices(deviceDefinitions: list[dict]):
 
 def validateConfiguration(deviceConfigs: list[dict]):
     deviceDict = {}
-    for device in deviceConfigs:
-
+    sorted_device_configs = sort_device_configs(deviceConfigs)
+    for device in sorted_device_configs:
         if deviceDict.get(device['id'], None) is None:
             deviceDict[device['id']] = device
         else:
             raise Exception("Duplicate device id {0}".format(device['id']))
+        if device.get('dependencies'):
+            for dependency in device["dependencies"]:
+                # Given that the device list has been sorted by dependencies,
+                # if one of our dependencies is not in the dict of checked
+                # devices, there must be a circular dependency.
+                if deviceDict.get(dependency, None) is None:
```

Comment:
So this is checking for a circular dependencies, but you're right that it's also checking for missing dependencies and thus will give inaccurate results. It's detecting circular dependencies because the sorting function used first compares "if a is a dependency of b then a < b", so in the simple case that both a depends on b and b depends on a, then the sorting order will be arbitrary, and then since we iterate through the list in order if we come across a configuration that has dependencies that have not already been seen, there must be a circular dependency somewhere. I just changed the code to use your implementation for this since it's more intuitive.

Reply:
Ahh I see now, very clever. Apologies I missed that.